src/main/java/reactor/netty/channel/MonoSendMany.java

```diff
@@ -358,38 +383,48 @@ void onInterruptionCleanup() {
 			if (queue == null) {
 				return;
 			}
-			Context context = null;
-			while (!queue.isEmpty()) {
-				I sourceMessage = queue.poll();
-				if (sourceMessage != null) {
-					parent.sourceCleanup.accept(sourceMessage);
-					if (context == null) {
-						context = actual.currentContext();
-					}
-					Operators.onDiscard(sourceMessage, context);
-				}
+
+			if (sourceMode == ASYNC) {
+				// delegates discarding to the queue holder to ensure there is no racing on draining from the SpScQueue
+				// fused queue should discard elements on clear
+				discardAsyncWithTermination(WIP, this, queue);
+			}
+			else {
+				Context context = currentContext();
+				discardWithTermination(WIP, this, queue, context);
 			}
 		}
 
 		boolean checkTerminated() {
 			return pending == 0 && terminalSignal != null;
 		}
 
-		void trySchedule(@Nullable Object data) {
-			if (WIP.getAndIncrement(this) == 0) {
-				try {
-					if (eventLoop.inEventLoop()) {
-						run();
-						return;
+		void trySchedule() {
+			int wip = wipIncrement(WIP, this);
+			if (wip != 0) {
+				if (wip == Integer.MIN_VALUE) {
+					if (sourceMode == ASYNC) {
+						queue.clear();
```

Comment:
can you expand on `clear()` vs discard-draining the queue in ASYNC mode ?

Reply:
We have that pattern in a reactor where the queue owner discards messages over `queue.clear()`. In the case of fusion, queue is a proxy subscription that refers to another queue in, e.g. UnicastProcessor or publishOn.

Reply:
https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/publisher/UnicastProcessor.java#L503

Reply:
ah sorry of course. `queue` is the `QueueSubscription` 🤦